libbeat/outputs/kafka/client.go

```diff
@@ -98,50 +105,99 @@ func (c *client) Connect() error {
 	c.wg.Add(2)
 	go c.successWorker(producer.Successes())
 	go c.errorWorker(producer.Errors())
+	go c.publishLoop()
 
 	return nil
 }
 
 func (c *client) Close() error {
 	debugf("closed kafka client")
 
+	stop <- true
 	c.producer.AsyncClose()
 	c.wg.Wait()
 	c.producer = nil
 	return nil
 }
 
+func (c *client) reconnect() error {
+	c.producer.AsyncClose()
+	c.wg.Wait()
+	c.producer = nil
+
+	producer, err := sarama.NewAsyncProducer(c.hosts, &c.config)
+	if err != nil {
+		logp.Err("Kafka connect fails with: %v", err)
+		return err
+	}
+
+	c.producer = producer
+
+	c.wg.Add(2)
+	go c.successWorker(producer.Successes())
+	go c.errorWorker(producer.Errors())
+	return nil
+}
+
 func (c *client) Publish(batch publisher.Batch) error {
 	events := batch.Events()
 	c.observer.NewBatch(len(events))
 
 	ref := &msgRef{
-		client: c,
-		count:  int32(len(events)),
-		total:  len(events),
-		failed: nil,
-		batch:  batch,
+		client:  c,
+		count:   int32(len(events)),
+		total:   len(events),
+		failed:  nil,
+		batch:   batch,
+		retry:   false,
+		backoff: common.NewBackoff(nil, 1*time.Second, 60*time.Second),
 	}
 
+	publish <- ref
+	return nil
+}
+
+func (c *client) publish(msgRef *msgRef) error {
+	if msgRef.retry {
+		c.reconnect()
+	}
```

Comment:
@urso, this is a re-work of https://github.com/elastic/beats/pull/7933 to provide fully asynchronous retries for any failed messages in batches sent to Kafka. I was testing it before I did a final cleanup on it and ran into this [Sarama issue](https://github.com/Shopify/sarama/issues/665). The call to `reconnect` above is a workaround with minimal performance impact during normal operation but more potential impact when retries have been queued up. I would be interested to know if you think we can salvage this approach, if you think a different approach will be necessary, or if you see a way to avoid this problem altogether?

Reply:
The original ticket says sarama blocks if Retry.Max is 0. We should set it at least to 1 :)

We can force libbeat to do the reconnect cycle by returning an error from `Publish`, if that is what we want. E.g. try a few times with circuite breaker. And if half-open state did fail n-times have `Publish` return an error.